### PR TITLE
Add context close codes and make extensibility binary

### LINKED ISSUE #53: Do context ID closures need more details?
In #52, we introduce the CLOSE_DATAGRAM_CONTEXT capsule which allows endpoints to close a context. That message currently only contains the context ID to close. @bemasc suggests that we may want to add more information there.
* Should we differentiate between "context close" and "context registration rejected"?
* Should we add an error code?

In particular, do folks have use cases for these features?

### PR DESCRIPTION
**Pull request by DavidSchinazi**
This PR adds a new concept of context close codes which
will allow applications that use contexts to indicate
the reason for closing a context.

This PR also makes the extensibility mechanism binary as
that is easier to implement.

Closes #53.

### CODE REVIEW

File: draft-ietf-masque-h3-datagram.md [hunk 1]
```diff
@@ -455,6 +442,74 @@ it is allowed to send an HTTP/3 datagram and how to process them from
...
+information.
+
+DENIED (code=0x01):
+
+: This indicates that the sender has rejected the context registration based on
+its local policy. The endpoint that had originally registered this context MUST
+NOT try to register another context with the same context extensions on this
+stream.
+
+RESSOURCE_LIMIT (code=0x02):
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:494]:
```suggestion
RESOURCE_LIMIT (code=0x02):
```

File: draft-ietf-masque-h3-datagram.md [hunk 2]
```diff
@@ -455,6 +442,74 @@ it is allowed to send an HTTP/3 datagram and how to process them from
...
+
+: This indicates that the context was closed to save resources. The recipient
+SHOULD limit its future registration of resource-incentive contexts.
+
+If CLOSE_CODE is not present in a CLOSE_DATAGRAM_CONTEXT capsule, the recipient
+SHALL treat it as if the NO_ERROR code was present. Close codes are registered
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:500]: Omitting the code seems like a micro optimization. Is there a strong case for this?

File: draft-ietf-masque-h3-datagram.md [hunk 3]
```diff
@@ -582,31 +637,80 @@ This registry initially contains the following entries:
...
+
+~~~
++------------------------------+-------+---------------+
+| Context Close Code           | Value | Specification |
++------------------------------+-------+---------------+
+| NO_ERROR                     | 0x00  | This Document |
++------------------------------+-------+---------------+
+| DENIED                       | 0x01  | This Document |
++------------------------------+-------+---------------+
+| RESSOURCE_LIMIT              | 0x02  | This Document |
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:711]:
```suggestion
| RESOURCE_LIMIT              | 0x02  | This Document |
```

File: draft-ietf-masque-h3-datagram.md [hunk 4]
```diff
@@ -731,7 +835,7 @@ DATAGRAM                       -------->
 STREAM(44): CAPSULE             -------->
   Capsule Type = REGISTER_DATAGRAM_CONTEXT
   Context ID = 2
-  Extension String = "ip=192.0.2.42,port=443"
+  Context Extension = {IP_COMPRESSION=192.0.2.42:443}
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:838]: Longer term, we might want to explain the presentation format for extensions but for now I think this is ok.

Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md:842]: Agreed, these example are likely to be removed before publication.

File: draft-ietf-masque-h3-datagram.md (hunk 2 above)
Comment (DavidSchinazi) [reply draft-ietf-masque-h3-datagram.md]: That's a good point, fixed.

**Review by LPardue — APPROVED** [review]
LGTM thanks

File: draft-ietf-masque-h3-datagram.md [hunk 5]
```diff
@@ -731,7 +839,7 @@ DATAGRAM                       -------->
...
```
Comment (bemasc) [inline draft-ietf-masque-h3-datagram.md:842]: I find this example kind of confusing, especially now that contexts are bidirectional.  A more realistic hypothetical might be something like `IP_COMPRESSION=tcp,192.0.2.41:54321,192.0.2.42:443`.  Or just `IP_COMPRESSION=<TBD connection info>`.

**Review by bemasc — APPROVED** [review]

**Pull request merged**